Make FileAndIndex class ordered and hashable

vepadulano opened the pull request:
The FileAndIndex class needs to be ordered and hashable because in `get_clustered_ranges` objects of this class are stored in a set and sorted to find, per each partition of the distributed RDataFrame, what files need to be considered.
Python classes get `__eq__` and `__hash__` methods by default, but not the ordering ones. Without this commit, `filelist` members of the `TreeRange` objects resulting from `get_clustered_ranges` would just be a 1:1 copy of the filenames that each cluster in the chain of that partition belongs to.
For example, given a tree with ten thousand entries and ten clusters split in two partitions, one would get:

Partition 0
Range of entries [0, 5000]
Filelist of the partition ['tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root']

Partition 1
Range of entries [5000, 10000]
Filelist of the partition ['tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root', 'tree10000entries10clusters.root']

So effectively we would be sending ranges with one filename per cluster in the range, which would lead to wrong reads.

vepadulano commented:
I found out about this while doing some tests for the PR that will remove the `Range` operation. We actually have tests for the machinery that creates the ranges. Unfortunately, they were all passing through the `rangesToTuples` function that effectively removed information from the `TreeRange` objects so the filelist was never checked.

vepadulano commented:
This time deepcode might be right with its warning:
```diff
def treeranges_to_tuples(ranges):
    """Convert TreeRange objects to tuples with the shape (start, end, filelist)"""
-    return list(map(lambda r: (r.start, r.end, r.filelist), ranges))
+    return [(r.start, r.end, r.filelist) for r in ranges]
```

eguiraud commented:
does this warrant a test?

vepadulano commented:
> does this warrant a test?

Yes definitely, I already modified the available tests to check the filelist. I can also add a new one to check a filelist with more than one file in it.

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/Ranges.py:
```diff
@@ -66,6 +67,21 @@ def __init__(self, filename, fileindex):
+        This type needs to be hashable to be part of a set in
+        get_clustered_ranges
+        """
+        return hash(self.filename)
```
could this make it harder to run twice on the same file?

vepadulano replied:
In general I wouldn't think so since the hashing is just to be able to put objects of this class in a Python `set`. The info that allows running twice on the same file is just the `fileindex` in the list of input files.

That being said, as soon as we switch to using `TEntryList` we won't be able to support running on the same file twice due to  https://github.com/root-project/root/issues/8505

vepadulano replied:
Anyway, if we want to be on the safe side, I could just make this `hash(self.filename) + hash(self.fileindex)`

eguiraud replied:
ah no ok, it's fine: `FileAndIndex` objects with same fileindex will also always have same filename

eguiraud commented on bindings/experimental/distrdf/test/test_buildranges.py:
```diff
@@ -208,42 +211,17 @@ def test_clustered_ranges_with_many_clusters_four_partitions(self):
-                                              range(step, end + 1, step))]
-
-        self.assertListEqual(ranges, ranges_reqd)
-
```
why is this test removed by the first commit in the PR?

vepadulano replied:
It's exactly the same as the following `test_buildranges_with_clustered_ranges` . Maybe at some point in the past they were testing different things, but I doubt it since `build_ranges` calls anyway `get_clustered_ranges`. Dataset, number of partitions etc. are all the same in the two tests so I removed one of them

eguiraud replied:
ok but then this removal doesn't have anything to do with the commit or the PR :smile:

vepadulano replied:
Ok I will revert this change

vepadulano commented:
I rebased commits, reverting the change that removed a test that is repeated twice (will do that in another PR). I also split the change of the class and the adaptation of tests in two commits. If the CI reports green I will merge

vepadulano commented:
Small typo in a test :smile: should be good now
